src/expr/node_trie_algorithm.cpp

```diff
@@ -0,0 +1,88 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Andrew Reynolds
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Algorithms for node tries
+ */
+
+#include "expr/node_trie_algorithm.h"
+
+namespace cvc5 {
+
+void nodeTriePathPairProcessInternal(const TNodeTrie* t1,
+                                     const TNodeTrie* t2,
+                                     size_t arity,
+                                     size_t depth,
+                                     NodeTriePathPairProcessCallback& ntpc)
+{
+  if (depth == arity)
+  {
+    // We are at the leaves. If we split the path, process the data.
+    if (t2 != nullptr)
+    {
+      ntpc.processData(t1->getData(), t2->getData());
+    }
+  }
+  else if (t2 == nullptr)
+  {
+    // we are exploring paths with a common prefix
+    if (depth < (arity - 1))
+    {
+      // continue exploring paths with common prefix, internal to each child
+      for (const std::pair<const TNode, TNodeTrie>& tt : t1->d_data)
+      {
+        nodeTriePathPairProcessInternal(
```

Comment:
I know that this is somewhat annoying, but could we implement this iteratively? I've run into issues with deep tries and recursive methods before. Given that this is a general util, I think that it would be good to make it work in those cases, too.